src/components/atoms/Timeline/index.tsx

```diff
@@ -0,0 +1,197 @@
+
+import { styled } from "@reearth/theme";
+
+export interface Props {
+  /**
+   * @description
+   * This value need to be epoch time.
+   */
+  currentTime: number;
+  /**
+   * @description
+   * These value need to be epoch time.
+   */
+  timelineRange: [start: number, end: number];
+  renderIcon: () => ReactElement;
+  iconWidth: number;
+}
+
+const EPOCH_SEC = 1000;
+const PADDING_HORIZONTAL = 12;
+const STRONG_MEMORY_HOURS = 3;
+const STRONG_MEMORY_WIDTH = 2;
+const NORMAL_MEMORY_WIDTH = 1;
+const GAP_HORIZONTAL = 14;
+const HOURS_SECS = 3600;
+const MEMORY_INTERVAL = 600;
+
+const Timeline: React.FC<Props> = ({ currentTime, timelineRange, renderIcon, iconWidth }) => {
+  const [startRange, endRange] = timelineRange;
+  const epochDiff = endRange - startRange;
+
+  // convert epoch diff to second.
+  const memoryCount = useMemo(
+    () => Math.trunc(epochDiff / EPOCH_SEC / MEMORY_INTERVAL),
+    [epochDiff],
+  );
+  const startRangeDate = useMemo(() => new Date(startRange), [startRange]);
+  const hoursCount = Math.trunc(HOURS_SECS / MEMORY_INTERVAL);
+  // Convert memory count to pixel.
+  const currentPosition = useMemo(() => {
+    const diff = Math.min((currentTime - startRange) / EPOCH_SEC / MEMORY_INTERVAL, memoryCount);
+    const strongMemoryCount = diff / (hoursCount * STRONG_MEMORY_HOURS);
+    return Math.max(
+      diff * GAP_HORIZONTAL +
+        (diff - strongMemoryCount) * NORMAL_MEMORY_WIDTH +
+        strongMemoryCount * STRONG_MEMORY_WIDTH +
+        STRONG_MEMORY_WIDTH / 2,
+      0,
+    );
+  }, [currentTime, startRange, memoryCount, hoursCount]);
+
+  return (
+    <Container>
+      <MemoryBox>
+        <MemoryList startRange={startRangeDate} memoryCount={memoryCount} hoursCount={hoursCount} />
+        <Icon
+          style={{
+            left: currentPosition + PADDING_HORIZONTAL - iconWidth / 2,
+          }}>
+          {renderIcon()}
+        </Icon>
+      </MemoryBox>
+    </Container>
+  );
+};
+
+const Container = styled.div`
+  background: ${({ theme }) => theme.main.lighterBg};
+  width: 100%;
+  display: flex;
+  align-items: center;
+  padding: 5px 12px;
+  box-sizing: border-box;
+`;
+
+const MemoryBox = styled.div`
+  border: 1px solid #b5b5b5;
+  border-radius: 5px;
+  box-sizing: border-box;
+  position: relative;
+  overflow-x: auto;
+  overflow-y: hidden;
+  width: 100%;
+  ::-webkit-scrollbar {
+    height: 5px;
+  }
+  ::-webkit-scrollbar-track {
+    background-color: transparent;
+  }
+  ::-webkit-scrollbar-thumb {
+    border-radius: 5px;
+    background-color: #aaa;
+  }
+`;
+
+const Icon = styled.div`
+  position: absolute;
+  top: 2px;
+`;
+
+const MONTH_LABEL_LIST = [
+  "Jan",
+  "Feb",
+  "Mar",
+  "Apr",
+  "May",
+  "Jun",
+  "Jul",
+  "Aug",
+  "Sep",
+  "Oct",
+  "Nov",
+  "Dec",
+];
+
+type MemoryListProps = {
+  startRange: Date;
+  memoryCount: number;
+  hoursCount: number;
+};
+
+const MemoryList: FC<MemoryListProps> = memo(
```

Comment:
I'd like to hear your thoughts on when to use memos; we didn't use memos much in reearth-web, but if we should try to use them more actively, we can consider memoizing other components.

Reply:
Basically, I think we need to use `memo` in some component like components that are rarely rendered. Especially atom components should be wrapped with `memo` because these component's props are rarely changed.
On the other hand, I think we should not wrap component that is often rendered because this is rarely memorized even if component is wrapped.